p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -440,6 +472,156 @@ impl P2pNetworkPubsubState {
+                            });
+                        }
+                    }
+                    GossipNetMessageV2::SnarkPoolDiff {
+                        message: NetworkPoolSnarkPoolDiffVersionedStableV2::AddSolvedWork(work),
+                        nonce,
+                    } => {
+                        dispatcher.push(P2pChannelsSnarkAction::Libp2pReceived {
+                            peer_id,
+                            snark: Box::new(work.1.into()),
+                            nonce: nonce.as_u32(),
+                        });
+                    }
+                    _ => {}
+                }
+
+                dispatcher.push(P2pNetworkPubsubAction::BroadcastValidatedMessage {
+                    message_id: super::BroadcastMessageId::MessageId { message_id },
+                });
+                Ok(())
+            }
+            P2pNetworkPubsubAction::BroadcastValidatedMessage { message_id } => {
+                let Some((message_id, message)) =
+                    pubsub_state.mcache.get_message_id_and_message(message_id)
+                else {
+                    return Ok(());
+                };
+
+                let peer_id = message.peer_id();
+                let raw_message = message.message().clone();
+                *message = P2pNetworkPubsubMessageCacheMessage::Validated {
+                    message: raw_message.clone(),
+                    peer_id,
+                    time: message.time(),
+                };
+                pubsub_state.reduce_incoming_validated_message(message_id, peer_id, &raw_message);
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+
+                Self::broadcast(dispatcher, state)
+            }
+            P2pNetworkPubsubAction::PruneMessages {} => {
+                let messages = pubsub_state
+                    .mcache
+                    .map
+                    .iter()
+                    .filter_map(|(message_id, message)| {
+                        if message.time() + Duration::from_secs(300) > time {
+                            Some(message_id.to_owned())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+
+                for message_id in messages {
+                    pubsub_state.mcache.remove_message(message_id);
+                }
+                Ok(())
+            }
+            P2pNetworkPubsubAction::RejectMessage {
+                message_id,
+                peer_id,
+                ..
+            } => {
+                let mut peer_id = peer_id;
+                if let Some(message_id) = message_id {
+                    let Some((_message_id, message)) =
+                        pubsub_state.mcache.get_message_id_and_message(message_id)
+                    else {
+                        return Ok(());
+                    };
+
+                    if peer_id.is_none() {
```

Comment:
is this correct? under which conditions is `peer_id` none? and can we really take the originator as the actual peer id in such cases?

Reply:
But when can that be `None`? I am confused about this switch between who sent us the message and the originator